tests/ui/unbound_return_lifetimes.rs

```diff
@@ -0,0 +1,74 @@
+#![allow(unused, dead_code, clippy::needless_lifetimes)]
+#![warn(clippy::unbound_return_lifetimes)]
+
+use std::collections::HashMap;
+use std::hash::Hash;
+
+struct FooStr(str);
+
+fn unbound_fun<'a>(x: impl AsRef<str> + 'a) -> &'a FooStr {
+    let x = x.as_ref();
+    unsafe { &*(x as *const str as *const FooStr) }
+}
+
+fn bound_fun<'a>(x: &'a str) -> &'a FooStr {
+    unsafe { &*(x as *const str as *const FooStr) }
+}
+
```

Comment:
Also perhaps some methods which return tuples or opaque types (impl trait) would be nice.